### PR TITLE
Serialization

### PR DESCRIPTION
**Pull request by tlepoint**
Serialization of the poly and poly_p using the (optional) cereal library (https://github.com/USCiLab/cereal) (if the library is not present, NFLlib still compiles and successfully tests).

I considered using boost rather than cereal, but either it requires to add boost as a dependency to add the serialize function (which I don't want), or I could not get the non intrusive version to work, possibly because the class does not expose "_data" to reconstruct the class state.

Note that:
- the master branch of cereal is not active but the development branch is;
- libcereal-dev is not in precise (Ubuntu), but is in xenial (Ubuntu 16.04LTS http://packages.ubuntu.com/xenial/libcereal-dev), so I had to use a git clone command 

What do you suggest?

1. We could not use cereal, but then I need help with boost.
2. Is there a way to use xenial in travis and to add libcereal-dev as a package in the .travis.yml file?
3. The current solution works well enough.

### CODE REVIEW

**Comment by carlosaguilarmelchor** [comment]
Nice work Tancrède. Have you tested the compactness/speed of cereal ? (to compare with what we would obtain by doing the serialization manually, as in practice you only save _data)

Correct me if I am wrong but imho the tests should :
- serialize p0 to Arch 
- copy p0 to p0'
- **free** p0 (to avoid false negatives on the tests due to pointers)
- déserialize from Arch p1
- test that p1 = p0'


Carlos

**Comment by serge-sans-paille** [comment]
hi tancrede. I don't have any personnal experience with cereal,. i did use boost.serialize in the past and it was ok. I also received good feedback on cereal lastly, so no strong opinion there.

Why do you need serialization by the way? i'll review your post later ;-)

**Comment by serge-sans-paille** [comment]
btw +1 for the tests @carlosaguilarmelchor proposes

**Comment by tlepoint** [comment]
I modified the test file as suggested.

I didn't try to do a manual serialization, I liked the C++ way of doing things with cereal, and I am not sure how to do it properly by hand. Do you think we should do it by hand?

I think serialization is a nice addition to the library (the overhead in term of code being minimal) and will be useful in applications (e.g. in [FV-NFLlib](https://github.com/CryptoExperts/FV-NFLlib), I would like to serialize ciphertexts, public/secret keys, etc.).

**Comment by carlosaguilarmelchor** [comment]
I suppose it is better to do it with cereal but as it may be used for sending ciphertexts over the network and have an influence on performance measures it would be nice to know that we are not paying too much for it.

**Comment by tlepoint** [comment]
On my computer (on battery), I have the following results

- 8_60_uint32_t

```
Cereal output serialization: 0.286 us
Cereal input serialization: 0.304 us
Manual output serialization: 0.196 us
Manual input serialization: 0.202 us
```

- 512_30_uint32_t

```
Cereal output serialization: 10.04 us
Cereal input serialization: 6.156 us
Manual output serialization: 4.518 us
Manual input serialization: 4.514 us
```

- 1024_60_uint32_t

```
Cereal output serialization: 32.526 us
Cereal input serialization: 31.886 us
Manual output serialization: 18.036 us
Manual input serialization: 17.256 us
```

- 8192_124_uint64_t

```
Cereal output serialization: 572.434 us
Cereal input serialization: 373.856 us
Manual output serialization: 354.85 us
Manual input serialization: 319.474 us
```

What do you think? We could use the manual functions, the code is eventually simple.

**Comment by aguinet** [comment]
The code looks fine for me, I agree with @carlosaguilarmelchor for the behavior of the tests ;)
I've got a question about your serailisation : you just need to output a bucnh of bytes, and "upper"l ayers know which type of polynom you have ?
If you indeed just have a fixed size array to write back and read forth, use your manual version :)

**Comment by tlepoint** [comment]
Just updated to use manual functions instead of cereal (since we need the parameters at compilation, it makes sense to only save _data). Is that OK with everyone?

**Comment by carlosaguilarmelchor** [comment]
@tlepoint saving only _data is perfectly ok with me.

I still have two questions tough. 

1) Did you run the tests for libcereal with a binary archive ? (it is the closest to the manual serialization and they say it is the fastest) If this is not the case can you rerun the tests with a binary *and* portable_binary archive ? 

2) Is the manual serialization compatible with libcereal and/or boost:serialize ? I mean, if there is an object O with a poly_t as an attribute and O is to complex to serialize manually will libcereal/boost be able to use the serialization functions we defined manually for the poly_t ?

Thx for the nice work btw :)

**Comment by tlepoint** [comment]
1) Yes I used cereal::BinaryOutputArchive and cereal::BinaryInputArchive
2) No it will not be compatible with libcereal in the sense you're suggesting. The object O will have to implement the serialize() function, but if it contains a polynomial then I think you'll have to reinterpret_cast the begin() as an array to save it with cereal because poly_t does not implement libcereal serialize function. That's why I wanted to use boost or libcereal in the first place.

File: include/nfl/poly.hpp [hunk 1]
```diff
@@ -151,6 +151,15 @@ class poly {
   void ntt_pow_phi() { base.ntt_pow_phi(*this);}
   void invntt_pow_invphi() { base.invntt_pow_invphi(*this); }
 
+  /* manual serializers
+  */
+  inline void serialize(std::ostream& outputstream) {
```
Comment (serge-sans-paille) [inline include/nfl/poly.hpp]: inline is redundant for inner classes. Serialize should be const. use _data.data() before the reinterpretcast looks safer to me.

Comment (serge-sans-paille) [inline include/nfl/poly.hpp]: inline is redundnat for inner classes. Serialize should be const. use _data.data() before the reinterpretcast looks safer to me.

File: include/nfl/poly.hpp [hunk 2]
```diff
@@ -151,6 +151,15 @@ class poly {
...
+    outputstream.write(reinterpret_cast<char*>(_data), N * sizeof(T));
+  }
+  inline void deserialize(std::istream& inputstream) {
+    inputstream.read(reinterpret_cast<char*>(_data), N * sizeof(T));
+  }
```
Comment (serge-sans-paille) [inline include/nfl/poly.hpp]: BTW, this kind of serialization is not portable in terms of « cross architecture portability ». It should be mentionned in the doc.

Comment (tlepoint) [inline include/nfl/poly.hpp]: Would boost/libcereal be portable? 

What should we do eventually? Do it manually, use boost (I couldn't get it to work with adding boost as a dependence), use libcereal? I was up for libcereal but then we discussed about doing it manually, and eventually it seems a problematic approach.

Comment (serge-sans-paille) [inline include/nfl/poly.hpp]: > Would boost/libcereal be portable? 

They can be (at least for boost) is you choose a portable backend,
generally a textual one

> What should we do eventually? Do it manually, use boost (I couldn't get it to work with adding boost as a dependence), use libcereal? I was up for libcereal but then we discussed about doing it manually, and eventually it seems a problematic approach.

It really depends on the use case, and I don't know about them.
basically if the data are meant to be shared over the network, then
using a serialization library would avoid this issues.

**Comment by carlosaguilarmelchor** [comment]
In fact my question was : "isn't there a way to do the manual serialization in a way that it will be reused by libcereal/boost if someone uses them for higher level objects ?"

For portability I don't see any other possibly issue than endianness. Given that little-endian is present on any platform in which we may imagine NFLlib used I think it is enough to do as @serge-sans-paille proposed: saying on the doc of the functions that serialization should not be used in the rare cases a big-endian machine will use NFLlib and communicate serialized data to a little-endian machine.

**Comment by tlepoint** [comment]
> In fact my question was : "isn't there a way to do the manual serialization in a way that it will be reused by libcereal/boost if someone uses them for higher level objects ?"

I don't think so.

**Comment by carlosaguilarmelchor** [comment]
What about : http://uscilab.github.io/cereal/archive_specialization.html ?
I have the feeling it is possible to overload what libcereal does and have all the benefits but I am unsure of how to do it.

**Comment by carlosaguilarmelchor** [comment]
+1 for merge

**Comment by tlepoint** [comment]
@serge-sans-paille @aguinet Is it OK for you?

File: tests/CMakeLists.txt [hunk 3]
```diff
@@ -1,21 +1,20 @@
...
-    #"1024, 37200, uint64_t"
-    "1024, 60, uint32_t"
-    #"1024, 3000, uint32_t"
-    "512, 30, uint32_t"
-	#"512, 14, uint16_t"
-	#"256, 14, uint16_t"
     "8, 60, uint32_t"
+    "128, 14, uint16_t"
+    "1024, 60, uint32_t"
+    "8192, 124, uint64_t"
```
Comment (serge-sans-paille) [inline tests/CMakeLists.txt]: Why a change in the values there? (512 / 30 disappears)

File: .travis.yml [hunk 4]
```diff
@@ -20,6 +20,7 @@ env:
     - NFL_CXX=clang++-3.5 NFL_OPTIMIZED=OFF
 
 script:
-    - cmake . -DNFL_OPTIMIZED=$NFL_OPTIMIZED -DCMAKE_CXX_COMPILER=$NFL_CXX -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=$PWD/install
+    - git clone https://github.com/USCiLab/cereal
```
Comment (serge-sans-paille) [inline .travis.yml]: cereal is header only?

Comment (tlepoint) [inline .travis.yml]: Yes!

File: tests/CMakeLists.txt (hunk 3 above)
Comment (tlepoint) [inline tests/CMakeLists.txt]: Well, we were not testing the 16-bit words - I think that was a mistake, so I replaced 512/30 by 128/14.

**Pull request merged**